source/common/thread_local/thread_local_impl.cc

```diff
@@ -17,93 +17,121 @@ thread_local InstanceImpl::ThreadLocalData InstanceImpl::thread_local_data_;
   ASSERT(std::this_thread::get_id() == main_thread_id_);
-  ASSERT(!shutdown_);
+  ASSERT(state_ != State::Shutdown);
 
   if (free_slot_indexes_.empty()) {
     SlotImplPtr slot(new SlotImpl(*this, slots_.size()));
-    auto wrapper = std::make_unique<Bookkeeper>(*this, std::move(slot));
-    slots_.push_back(wrapper->slot_.get());
-    return wrapper;
+    slots_.push_back(slot.get());
+    return slot;
   }
   const uint32_t idx = free_slot_indexes_.front();
   free_slot_indexes_.pop_front();
   ASSERT(idx < slots_.size());
   SlotImplPtr slot(new SlotImpl(*this, idx));
   slots_[idx] = slot.get();
-  return std::make_unique<Bookkeeper>(*this, std::move(slot));
+  return slot;
+}
+
+InstanceImpl::SlotImpl::SlotImpl(InstanceImpl& parent, uint64_t index)
+    : parent_(parent), index_(index), still_alive_guard_(std::make_shared<bool>()),
+      active_callback_guard_(
+          new bool,
+          [&parent, still_alive_guard = std::weak_ptr<bool>(still_alive_guard_), this](bool* b) {
+            // If TLS is in the running state, signal that this slot is ready to be destroyed. This
+            // may happen on a worker when callbacks are run. We do not do this check during:
+            // a) initializing: there are error flows during initialization in which it is too
+            //    complicated to destroy all pending callbacks before the slot is destroyed. Given
+            //    that no workers are running there is nothing to synchronize anyway.
+            // b) shutting down: no slots will be destroyed between the time shutdown starts and all
+            //    workers exit, so there is nothing to synchronize once all workers have exited.
+            //    Additionally, there are initialization failure cases in which we move directly
+            //    from initializing to shutting down, so for similar reasons to (a) we don't
+            //    synchronize.
+            // c) when we can't lock the still alive guard. It's only possible for that to happen
+            //    when a slot was destroyed before we entered the running state (during xDS init).
+            //    This is guaranteed by the blocking in slot destruction. Thus, no further
+            //    synchronization is needed.
+            if (parent.state_ == State::Running && still_alive_guard.lock()) {
+              absl::MutexLock lock(&shutdown_mutex_);
+              ready_to_destroy_ = true;
+            }
+
+            delete b;
+          }) {}
+
+InstanceImpl::SlotImpl::~SlotImpl() {
+  // Reset active_callback_guard_ so the only active references will be held by pending callbacks.
+  active_callback_guard_.reset();
+  // If running only, synchronize slot removal with all callbacks being drained from workers.
+  // See the constructor for why we only do this in the running state.
+  if (parent_.state_ == State::Running) {
+    absl::MutexLock lock(&shutdown_mutex_);
+    shutdown_mutex_.Await(absl::Condition(&ready_to_destroy_));
```

Comment:
is there a deadlock risk if a worker thread accidentally posts the wrapped callback back to main thread?

Reply:
In practice no, since the worker can't actually get access to the wrapped callback, only the unwrapped one.

Reply:
I see.
Another risk though a long event in worker thread may block the main thread for a long while.
 
can we combine this active checking together with the "deferred deletion" trick?

I actually dont know why the bookkeeper trick would fail in the first place. that trick tries to guarantee the liveness of the slot when there are flying cb in worker threads as well.

Reply:
> Another risk though a long event in worker thread may block the main thread for a long while.

Yes, this is possible. In practice, I'm not sure this matters since slot deletion is a rare operation and the workers should not be blocking, and we also have deadlock detection.

> can we combine this active checking together with the "deferred deletion" trick?

What do you mean?

> I actually dont know why the bookkeeper trick would fail in the first place. that trick tries to guarantee the liveness of the slot when there are flying cb in worker threads as well.

I explained this here: https://github.com/envoyproxy/envoy/issues/12364#issuecomment-675631096. Bookkeeper does not fix the issue in which people capture something in the lambda, and that thing gets destroyed before the callback runs. This is the bug in the linked issue and have had this bug at least 2-3 times before, it comes up quite often. With that said, it's possible we could get a long way with something like this: https://github.com/mesos/clang-tools-extra/blob/mesos_90/clang-tidy/mesos/ThisCaptureCheck.cpp which @jpeach sent me.

Reply:
> > Another risk though a long event in worker thread may block the main thread for a long while.
> 
> Yes, this is possible. In practice, I'm not sure this matters since slot deletion is a rare operation and the workers should not be blocking, and we also have deadlock detection.
> 
I have some concern here, the lock may take enough long to impact XDS or other OP on main thread, whe we have dozens of workers. for example, a RDS Config provider holds a tls slot, in SRDS we dynamically add or delete the 
RDS providers in each SRDS reponse. This could potentially put a performance hit on main thread.

> > can we combine this active checking together with the "deferred deletion" trick?
> 
> What do you mean?
> 
I mean, instead of hanging main thread, we could potentially put this slot into a deferred deletion queue, and schedule a deletionCB in main thread to cleanup all the slot instance when the ready_to_destroy_ bit flipped to true.

> > I actually dont know why the bookkeeper trick would fail in the first place. that trick tries to guarantee the liveness of the slot when there are flying cb in worker threads as well.
> 
> I explained this here: [#12364 (comment)](https://github.com/envoyproxy/envoy/issues/12364#issuecomment-675631096). Bookkeeper does not fix the issue in which people capture something in the lambda, and that thing gets destroyed before the callback runs. This is the bug in the linked issue and have had this bug at least 2-3 times before, it comes up quite often. With that said, it's possible we could get a long way with something like this: https://github.com/mesos/clang-tools-extra/blob/mesos_90/clang-tidy/mesos/ThisCaptureCheck.cpp which @jpeach sent me.

Reply:
> I mean, instead of hanging main thread, we could potentially put this slot into a deferred deletion queue, and schedule a deletionCB in main thread to cleanup all the slot instance when the ready_to_destroy_ bit flipped to true.

As I just explained to you in DM, this doesn't actually fix the problem. We really have only 2 options:
1) We have to prevent the problem (blocking)
2) We have to fix all call sites (probably clang-tidy plugin, etc.)

Reply:
My understanding is that the crash is only possible when worker has not started, 
But the cost at conditional wait is applied to working running state?! `parent_.state_ == State::Running`

Reply:
It's not just before workers have started. The issue also happens post worker start if a listener fails to get fully constructed for example during LDS update.

Reply:
My main suggestion here is that, in case something goes wrong (deadlock or just slow exit), use `AwaitWithTimeout` in a loop, logging `info` every N seconds so that an operator knows why we're stuck even if we don't have debug signals. Ideally we could log progress toward being able to move on.

In the best case that log gets printed very few times, and in the worst case it's very noisy but we'll have an easier time diagnosing.

Reply:
Yeah, I did consider also just always doing a timeout to make sure we don't deadlock, or do what you suggest with a limit.

Reply:
> It's not just before workers have started. The issue also happens post worker start if a listener fails to get fully constructed for example during LDS update.

Right, anytime a slot is destroyed before the worker thread play with it.

Can we enforce removing the slot destroy *after* worker thread drop reference? E.g. always wrap an "attempt to delete function" posted to master thread regardless user has the `complete_cb`?

Reply:
> Can we enforce removing the slot destroy after worker thread drop reference? E.g. always wrap an "attempt to delete function" posted to master thread regardless user has the complete_cb?

This is the way it works today. :) This is "bookkeeper". The problem is this method does not actually fix the dangling lambda capture issue. You simply can't fix it without blocking.

So like I said above, I think we have to decide:
1) Blocking is OK
2) Blocking is not OK and we should look into clang-tidy plugins or similar.

Reply:
Is mismanagement of object lifetimes the cause of the underlying bugs which this PR is trying to fix? If so, I'm not sure adding code complexity and blocking to paper over the issue is a good idea. I'd be much more in favor of seeing how far clang-tidy plugins get us.

Reply:
> Is mismanagement of object lifetimes the cause of the underlying bugs which this PR is trying to fix?

Yes. Though, to be fair, the existing code already does this (see Bookkeeper).

> If so, I'm not sure adding code complexity and blocking to paper over the issue is a good idea. I'd be much more in favor of seeing how far clang-tidy plugins get us.

Yes that's a fair viewpoint and I'm fine if that is the group consensus.

Reply:
I probably would have voted against bookkeeper too, FWIW. In general, if a user is doing something incorrectly, I'd rather we surface that information to the user so they can implement a fix on their end that makes sense, rather than trying to detect that they made a mistake and prevent it from causing problems. The latter feels more bug-prone to me.

Reply:
@ahedberg at a high level I agree with you. I'm torn on this one though, as this is a really thorny issue that has been the cause of many different crashes. The issue here is that lifetime alone is not necessarily a complete fix. Meaning, depending on how someone implements the captures, there can still be race conditions with deletion. Let me summarize the current state opinions below and we can think about a path forward.

Reply:
see https://github.com/envoyproxy/envoy/issues/7902 we will need either the bookkeeper or the change in this PR to fix that issue.